### PR TITLE
Fix bug in passing parameters to the python action

### PR DESCRIPTION
**Pull request by ioana-blue**
Fixed bug and added test to chec a couple of jsons, one with null

Subsumes #906

### CODE REVIEW

**Comment by ioana-blue** [comment]
This is ready for review & merge.

**Comment by rabbah** [comment]
I'm not sure that we need a complex test for null - here's a suggested alternative:

https://github.com/openwhisk/openwhisk/compare/master...rabbah:pr-906?expand=1

File: core/pythonAction/pythonaction.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def run():
     result = None
     try:
         exec(flask.g, namespace)
```
Comment (rabbah) [inline core/pythonAction/pythonaction.py]: why isn't this part of init: `exec(flask.g, namespace)` - wouldn't that save the eval overhead on every run?

Comment (ioana-blue) [inline core/pythonAction/pythonaction.py]: I think this evolved. At some point there was an eval part of init. I'll fix it part of this PR.

**Comment by ioana-blue** [comment]
One of your links above doesn't work for me, but I get what you're saying. Fixing.

File: core/pythonAction/pythonaction.py (hunk 1 above)
Comment (ioana-blue) [inline core/pythonAction/pythonaction.py]: Actually, I think this statement in particular is not needed (subsumed by the following exec), which is really a bug and not only an optimization issue. Good catch!

**Comment by ioana-blue** [comment]
@rr I addressed your comments and fixes are in place. Please take a careful look since I fixed another issue with providing feedback during init for when compilation fails (I think there was an issue opened for this, maybe from Nick). moving this to review.

File: tests/src/actionContainers/PythonActionContainerTests.scala [hunk 2]
```diff
@@ -73,6 +72,30 @@ class PythonActionContainerTests extends FlatSpec
...
+            val (initCode, _) = c.init(initPayload(code))
+            initCode should be(200)
+
+            val argss = List(
+                JsObject("user" -> JsNull),
+                JsObject("user" -> JsString("Momo")))
+
+            for (args <- argss) {
+                val (runCode, out) = c.run(runPayload(args))
+                println(out)
```
Comment (rabbah) [inline tests/src/actionContainers/PythonActionContainerTests.scala]: remove println.

File: core/pythonAction/pythonaction.py [hunk 3]
```diff
@@ -35,8 +35,16 @@ def init():
...
-        # store the code
-        flask.g = message["code"]
+        # store the compiled code
+        try:
+            flask.g = compile(message["code"], filename = '<string>', mode = 'exec')
+        except Exception:
+            flask.g = None
+            traceback.print_exc(file = sys.stderr)
+            sys.stderr.flush()
+            response = flask.jsonify({"error": "the action failed to compile. See logs for details." })
```
Comment (rabbah) [inline core/pythonAction/pythonaction.py]: the -> The (since you capitalized elsewhere).

File: core/pythonAction/pythonaction.py [hunk 4]
```diff
@@ -56,12 +64,18 @@ def run():
     if not isinstance(value, dict):
         flask.abort(403)
 
+    if flask.g is None:
+        # no code to execute
+        response = flask.jsonify({"error": "No code to execute (init failed). See logs for details." })
```
Comment (rabbah) [inline core/pythonAction/pythonaction.py]: init -> compilation? since you called it compilation above.

Comment (rabbah) [inline core/pythonAction/pythonaction.py]: this should never happen btw (ie if you're called when init failed, there's an internal system error).

**Comment by rabbah** [comment]
Can you link to the issue mentioned above (in the commit and this pr).
PG?

**Comment by rabbah** [comment]
Not sure if you meant this one: https://github.com/openwhisk/openwhisk/issues/493...?

File: core/pythonAction/pythonaction.py [hunk 5]
```diff
@@ -73,7 +87,7 @@ def run():
         response.status_code = 200
         return response
     else:
-        response = flask.jsonify({ "error": "the action did not return a dictionary", "action_output": result })
+        response = flask.jsonify({"error": "The action did not return a dictionary", "action_output": result })
```
Comment (rabbah) [inline core/pythonAction/pythonaction.py]: missing period.
